pytest-embedded-wokwi/pyproject.toml

```
@@ -0,0 +1,106 @@
+[build-system]
+requires = ["flit_core >=3.2,<4"]
+build-backend = "flit_core.buildapi"
+
+[project]
+name = "pytest-embedded-wokwi"
+authors = [
+    {name = "Fu Hanxi", email = "fuhanxi@espressif.com"},
+]
+readme = "README.md"
+license = {file = "LICENSE"}
+classifiers = [
+    "Development Status :: 5 - Production/Stable",
+    "Framework :: Pytest",
+    "Intended Audience :: Developers",
+    "License :: OSI Approved :: MIT License",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python :: 3 :: Only",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python",
+    "Topic :: Software Development :: Testing",
+]
+dynamic = ["version", "description"]
+requires-python = ">=3.7"
+
+dependencies = [
+    "pytest-embedded~=1.3.5",
+    "toml~=0.10.2",
```

```suggestion
    "toml; python_version < '3.11'",
```

Since python 3.11, the `tomllib` is natively supported.

```python
def load_toml(filepath: t.Union[str, Path]) -> dict:
    try:
        import tomllib  # python 3.11

        try:
            with open(str(filepath), 'rb') as fr:
                return tomllib.load(fr)
        except Exception as e:
            raise InvalidTomlError(filepath, str(e))
    except ImportError:
        import toml

        try:
            return toml.load(str(filepath))
        except Exception as e:
            raise InvalidTomlError(filepath, str(e))
```

---

Thanks for the suggestion, however, I see that `tomllib` only includes the `load()` and `loads()` methods, but there is no way to write toml files. Any suggestions how to go about this?

---

ah, sorry I didn't notice the call to the `write()`. The `tomllib` is only used for parsing. Let's keep it as what it is right now.

---

Thanks for reviewing my PR!

Regarding esp-idf - is there currently a way to use pytest embedded without esp-idf? I see there's also an Arduino plugin. We can possibly support the other use cases as well, I'm just not sure if it's worth the effort right now. WDYT?

Regarding `IdfUnityDutMixin` - I think this already works? We're adding the mix in [here](https://github.com/espressif/pytest-embedded/pull/230/files#diff-b5a6b0a6e20cb69a3ce927b538961a62b3a189870fb359992536e7cc2618cfa5R1228-R1231). Not sure if this is the best way to go about this, I just copied from the qemu case.

---

Yes, there is. For example, you may use `qemu` service alone, or `serial` service alone, to interact with a Qemu process, or with a `serial` object.

Regarding the `IdfUnityDutMixin`, yes, you've added the mixin when `idf` service is enabled. The situation here is slightly different. Unlike `qemu`, `wokwi` shouldn't be run alone, so inherit the mixin directly would more intuitive. This was my original idea.

Thanks for reminding me, if we're going to support arduino in the same package as well, I agree to use the name `pytest-embedded-wokwi`. We only need to rename the `WokwiCLI` to `WokwiIdfCLI` or something, to make the hard-coded `create_wokwi_toml` to IDF-specific.

So, in short... considering the future support to arduino, here's my opinions
- keep the current package name
- limit the service `wokwi`, it should be used together with `idf` (or `arduino` in the future)

  ```python
            if 'wokwi' in _services:
                from pytest_embedded_wokwi import WokwiDut

                classes[fixture] = WokwiDut
                kwargs[fixture].update(
                    {
                        'wokwi': None,
                    }
                )

                if 'idf' in _services:
                    from pytest_embedded_idf.unity_tester import IdfUnityDutMixin

                    mixins[fixture].append(IdfUnityDutMixin)
                else:
                    # - you may also declare a new type of error as well
                    # - or set the `idf` as a default fallback, if not exists
                    raise SystemExit('wokwi service should be used together with idf service')
  ```

- put the current `create_wokwi_toml` in IDF-specific class, use this class to initialize the `wokwi_cli` fixture when `idf` service is enabled. (we may use something else when `arduino` service is enabled in the future)

---

Thanks, I will look into doing these changes tomorrow. Other than `idf` and `arduino`, are there any other plugins planned?

What wokwi cli essentially needs is a list of bin files, their offsets, and and elf file if exists. If there is a common way to get the list of bin files and their offset, perhaps we can make it framework agnostic?

---

Thanks again for the detailed feedback, I pushed an update limiting `wokwi` to be used together with `idf`, as well as moving the IDF specific logic into a separate class.